### PR TITLE
fix(irq): prevent interrupt reservation and improve error handling

### PR DESCRIPTION
**Pull request by Diogo21Costa**
## PR Description

This PR enhances the interrupt reservation functions to prevent race conditions during the reservation process. It comprises the following changes:

1. Added a spinlock to ensure synchronized access to the critical section, preventing race conditions. The original code lacked this synchronization, potentially leading to issues. The spinlock now ensures only one thread can access the critical section at a time, resolving the race condition (https://github.com/bao-project/bao-hypervisor/commit/196a404e58dac8ba93f08b24b06360abe5f39a8e);
2. Introduced a boolean return value to indicate the success of interrupt reservation. This provides a more detailed assessment of interrupt assignment status, aiding in error handling (https://github.com/bao-project/bao-hypervisor/commit/fc2029b8e9dc294ef4e2bbfb6946fb0eaa42ad27);
3. Improved the overall error handling for interrupt assignment (https://github.com/bao-project/bao-hypervisor/commit/5887601d39b511b9a314e21ce77bf47ebda9ee68)

Your feedback on these changes is greatly appreciated.

### CODE REVIEW

File: src/core/interrupts.c [hunk 1]
```diff
@@ -12,6 +12,7 @@
 
 BITMAP_ALLOC(hyp_interrupt_bitmap, MAX_INTERRUPTS);
 BITMAP_ALLOC(global_interrupt_bitmap, MAX_INTERRUPTS);
+spinlock_t irq_reserve_lock;
```
Comment (josecm) [inline src/core/interrupts.c:15]: This lock is never initialized.

```suggestion
spinlock_t irq_reserve_lock = SPINLOCK_INITVAL;
```

File: src/arch/armv8/gic.c [hunk 2]
```diff
@@ -68,8 +68,10 @@ void gicd_init()
 
     /* No need to setup gicd->NSACR as all interrupts are  setup to group 1 */
 
-    interrupts_reserve(platform.arch.gic.maintenance_id,
-                       gic_maintenance_handler);
+    if(!interrupts_reserve(platform.arch.gic.maintenance_id,
+                       gic_maintenance_handler)) {
+        ERROR("Failed to assign GIC maintenance interrupt");
```
Comment (josecm) [inline src/arch/armv8/gic.c:73]:
```suggestion
        ERROR("Failed to reserve GIC maintenance interrupt");
```

File: src/arch/riscv/iommu.c [hunk 3]
```diff
@@ -334,7 +334,10 @@ void rv_iommu_init(void)
     rv_iommu.hw.reg_ptr->fqh = 0;
 
     // Allocate IRQ for FQ
-    interrupts_reserve(platform.arch.iommu.fq_irq_id, rv_iommu_fq_irq_handler);
+    if(!interrupts_reserve(platform.arch.iommu.fq_irq_id, rv_iommu_fq_irq_handler)) {
+        ERROR("Failed to assign IOMMU FQ interrupt");
```
Comment (josecm) [inline src/arch/riscv/iommu.c:338]:
```suggestion
        ERROR("Failed to reserve IOMMU FQ interrupt");
```

File: src/arch/riscv/sbi.c [hunk 4]
```diff
@@ -476,5 +476,7 @@ void sbi_init()
         }
     }
 
-    interrupts_reserve(TIMR_INT_ID, sbi_timer_irq_handler);
+    if(!interrupts_reserve(TIMR_INT_ID, sbi_timer_irq_handler)) {
+        ERROR("Failed to assign SBI TIMR_INT_ID interrupt");
```
Comment (josecm) [inline src/arch/riscv/sbi.c:480]:
```suggestion
        ERROR("Failed to reserve SBI TIMR_INT_ID interrupt");
```

File: src/core/interrupts.c [hunk 5]
```diff
@@ -40,7 +41,9 @@ inline void interrupts_init()
     interrupts_arch_init();
 
     if (cpu()->id == CPU_MASTER) {
-        interrupts_reserve(IPI_CPU_MSG, cpu_msg_handler);
+        if(!interrupts_reserve(IPI_CPU_MSG, cpu_msg_handler)) {
+            ERROR("Failed to assign IPI_CPU_MSG interrupt");
```
Comment (josecm) [inline src/core/interrupts.c:45]:
```suggestion
            ERROR("Failed to reserve IPI_CPU_MSG interrupt");
```

**Review by josecm — CHANGES_REQUESTED** [review]

File: src/core/interrupts.c (hunk 1 above)
Comment (Diogo21Costa) [reply src/core/interrupts.c]: That's a good point. Addressed in https://github.com/bao-project/bao-hypervisor/pull/97/commits/1a898b49d4bac726ca9a99e70feeb4affeb32ef0. Thank you for your feedback!

File: src/arch/armv8/gic.c (hunk 2 above)
Comment (Diogo21Costa) [reply src/arch/armv8/gic.c]: Thank you, included in https://github.com/bao-project/bao-hypervisor/pull/97/commits/878ad2b2ed85b8ed383a63e32650ee9363c624d8.

File: src/arch/riscv/iommu.c (hunk 3 above)
Comment (Diogo21Costa) [reply src/arch/riscv/iommu.c]: Thank you, fixed in https://github.com/bao-project/bao-hypervisor/pull/97/commits/878ad2b2ed85b8ed383a63e32650ee9363c624d8.

File: src/arch/riscv/sbi.c (hunk 4 above)
Comment (Diogo21Costa) [reply src/arch/riscv/sbi.c]: Thank you, fixed in https://github.com/bao-project/bao-hypervisor/pull/97/commits/878ad2b2ed85b8ed383a63e32650ee9363c624d8.

File: src/core/interrupts.c (hunk 5 above)
Comment (Diogo21Costa) [reply src/core/interrupts.c]: Thank you, fixed in https://github.com/bao-project/bao-hypervisor/pull/97/commits/878ad2b2ed85b8ed383a63e32650ee9363c624d8.

**Comment by Diogo21Costa** [comment]
@josecm consolidated all error message updates into a single commit (https://github.com/bao-project/bao-hypervisor/pull/97/commits/878ad2b2ed85b8ed383a63e32650ee9363c624d8). As this commit fixes the content introduced in https://github.com/bao-project/bao-hypervisor/pull/97/commits/5887601d39b511b9a314e21ce77bf47ebda9ee68, may I squash both commits to keep the git history clean?

The same applies to commits https://github.com/bao-project/bao-hypervisor/pull/97/commits/1a898b49d4bac726ca9a99e70feeb4affeb32ef0 and https://github.com/bao-project/bao-hypervisor/pull/97/commits/196a404e58dac8ba93f08b24b06360abe5f39a8e ( The first one introduces the spinlock initialization). Should they be squashed?

**Comment by josecm** [comment]
@Diogo21Costa Yeah, I'd say the history should be re-written and include the new fixes directly on the original commits.

**Comment by Diogo21Costa** [comment]
> @Diogo21Costa Yeah, I'd say the history should be re-written and include the new fixes directly on the original commits.

@josecm Done, thank you!

**Review by josecm — APPROVED** [review]

**Review by DavidMCerdeira — CHANGES_REQUESTED** [review]
This commit's description df3168ce3175a00bcc9bfeaa0a3000e0a04e0aff does not match the code changes

Otherwise seems good

**Comment by Diogo21Costa** [comment]
> This commit's description [df3168c](https://github.com/bao-project/bao-hypervisor/commit/df3168ce3175a00bcc9bfeaa0a3000e0a04e0aff) does not match the code changes
> 
> Otherwise seems good

@DavidMCerdeira Nice catch, thank you. Solved.

**Review by DavidMCerdeira — APPROVED** [review]

**Pull request merged**